filters/replicateto/replicateto.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2015 Zalando SE
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package replicateto
+
+import (
+	"fmt"
+	"github.com/zalando/skipper/filters"
+	"net/http"
+	"net/url"
+	"regexp"
+)
+
+const (
+	Name = "replicateTo"
+)
+
+type replicateToSpec struct{}
+
+type replicateType int
+
+const (
+	asBackend replicateType = iota + 1
+	pathModified
+)
+
+type replicateTo struct {
+	typ         replicateType
+	host        string
+	scheme      string
+	rx          *regexp.Regexp
+	replacement []byte
+}
+
+// Returns a new replication filter Spec, whose instances execute
+// exactly same Request against a shadow backend.
+// parameters: shadow backend url, optional - the expression to match and the replacement string.
+// Name: "replicateTo".
+
+// Example
+// Path("http://example.com/api/v3") -> replicateTo("https://api.example.com") -> "http://example.org/"
+//This route wil send incoming query to http://example.org/api/v3 but will also send
+// a copy of the query to https://api.example.com/api/v3. Note that scheme is changed
+
+// Example
+// Path("http://example.com/api/v3") -> replicateTo("https://api.example.com", ".*", "/v1/" ) -> "http://example.org/"
+//This route wil send incoming query to http://example.org/api/v3 but will also send
+// a copy of the query to https://api.example.com/v1/ . Note that scheme and path are changed
+
+func NewReplicateTo() *replicateToSpec {
+	return &replicateToSpec{}
+}
+
+//We do not touch response at all
+func (r *replicateTo) Response(filters.FilterContext) {}
+
+// Request is copied and then modified to adopt changes in new backend
+func (r *replicateTo) Request(fc filters.FilterContext) {
+	copyOfRequest := modifyRequest(r, fc.Request())
+	go http.DefaultClient.Do(&copyOfRequest)
+}
+
+// copies requests changes URL and Host in request.
+// If 2nd and 3rd params are given path is also modified by applying regexp
+func modifyRequest(rep *replicateTo, req *http.Request) http.Request {
```

Comment:
besides: this is not enough, because this way the two outgoing requests share the body. This may not be completely straightfoward to do without blocking and excessive buffering. Please have a look at using io.Pipe and io.TeeReader (see some good naming just there :) ). We can try to do it together some time.